miniapps/elasticity/operators/elasticity_operator.hpp

```diff
@@ -0,0 +1,305 @@
+   /// point rather than recompute the action.
+   bool use_cache_ = true;
+   mutable bool recompute_cache_ = false;
+   Vector dsigma_cache_;
+
+   /**
+    * @brief Wrapper for the application of the residual R(U).
+    *
+    * The wrapper is used in SetMaterial to instantiate the chosen kernel and
+    * erase the material type kernel. This is purely an interface design choice
+    * and could be replaced by an abstract base class for the material including
+    * virtual function calls.
+    */
+   std::function<void(const int, const Array<double> &, const Array<double> &,
+                      const Array<double> &, const Vector &, const Vector &,
+                      const Vector &, Vector &)>
+   element_apply_kernel_wrapper;
+
+   /**
+    * @brief Wrapper for the application of the gradient of the residual
+    *
+    *  K(U) dX = dR(U)/dU dX
+    */
+   std::function<void(const int, const Array<double> &, const Array<double> &,
+                      const Array<double> &, const Vector &, const Vector &,
+                      const Vector &, Vector &, const Vector &)>
+   element_apply_gradient_kernel_wrapper;
+
+   /**
+    * @brief Wrapper for the assembly of the gradient on each diagonal element
+    *
+    * Ke_ii(U) = dRe_ii(U)/dU
+    */
+   std::function<void(const int, const Array<double> &, const Array<double> &,
+                      const Array<double> &, const Vector &, const Vector &,
+                      const Vector &, Vector &)>
+   element_kernel_assemble_diagonal_wrapper;
+
+   /**
+    * @brief Set the material type.
+    *
+    * This method sets the material type by instantiating the kernels with a
+    * material_type object.
+    *
+    * @tparam material_type
+    * @param[in] material
+    */
+   template <typename material_type>
+   void SetMaterial(const material_type &material)
+   {
+      if (dim_ != 3)
+      {
+         MFEM_ABORT("dim != 3 not implemented");
+      }
+
+      element_apply_kernel_wrapper =
+         [=](const int ne, const Array<double> &B_, const Array<double> &G_,
+             const Array<double> &W_, const Vector &Jacobian_,
+             const Vector &detJ_, const Vector &X_, Vector &Y_)
+      {
+         const int id = (d1d_ << 4) | q1d_;
+         switch (id)
+         {
+            case 0x22:
+            {
+               ElasticityKernels::Apply3D<2, 2, material_type>(
+                  ne, B_, G_, W_, Jacobian_, detJ_, X_, Y_, material);
+               break;
+            }
+            case 0x33:
+            {
+               ElasticityKernels::Apply3D<3, 3, material_type>(
+                  ne, B_, G_, W_, Jacobian_, detJ_, X_, Y_, material);
+               break;
+            }
+            case 0x44:
+               ElasticityKernels::Apply3D<4, 4, material_type>(
+                  ne, B_, G_, W_, Jacobian_, detJ_, X_, Y_, material);
+               break;
+            default:
+               MFEM_ABORT("Not implemented: " << std::hex << id << std::dec);
```

Comment:
This should have a more descriptive message of the rejected configuration. E.g. `num dofs `d = {}, quad points 1d = {}`